presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/PrimitiveTypeMapBuilder.java

```diff
@@ -0,0 +1,86 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.iceberg;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import io.prestosql.spi.type.ArrayType;
+import io.prestosql.spi.type.MapType;
+import io.prestosql.spi.type.RowType;
+import io.prestosql.spi.type.Type;
+
+import java.util.List;
+import java.util.Map;
+
+import static io.prestosql.spi.type.StandardTypes.ARRAY;
+import static io.prestosql.spi.type.StandardTypes.MAP;
+import static io.prestosql.spi.type.StandardTypes.ROW;
+
+public class PrimitiveTypeMapBuilder
+{
+    private final ImmutableMap.Builder<List<String>, Type> builder = ImmutableMap.builder();
+
+    private PrimitiveTypeMapBuilder() {}
+
+    public static Map<List<String>, Type> makeTypeMap(List<Type> types, List<String> columnNames)
+    {
+        return new PrimitiveTypeMapBuilder().buildTypeMap(types, columnNames);
+    }
+
+    public Map<List<String>, Type> buildTypeMap(List<Type> types, List<String> columnNames)
+    {
+        for (int i = 0; i < types.size(); i++) {
+            visitType(types.get(i), columnNames.get(i), ImmutableList.of());
+        }
+        return builder.build();
+    }
+
+    private void visitType(Type type, String name, List<String> parent)
+    {
+        if (ROW.equals(type.getTypeSignature().getBase())) {
+            visitRowType((RowType) type, name, parent);
+        }
+        else if (MAP.equals(type.getTypeSignature().getBase())) {
+            visitMapType((MapType) type, name, parent);
+        }
+        else if (ARRAY.equals(type.getTypeSignature().getBase())) {
+            visitArrayType((ArrayType) type, name, parent);
+        }
+        else {
+            builder.put(ImmutableList.<String>builder().addAll(parent).add(name).build(), type);
+        }
+    }
+
+    private void visitArrayType(ArrayType type, String name, List<String> parent)
+    {
+        parent = ImmutableList.<String>builder().addAll(parent).add(name).add("array").build();
+        visitType(type.getElementType(), "array", parent);
+    }
+
+    private void visitMapType(MapType type, String name, List<String> parent)
+    {
+        parent = ImmutableList.<String>builder().addAll(parent).add(name).add("map").build();
+        visitType(type.getKeyType(), "key", parent);
+        visitType(type.getValueType(), "value", parent);
+    }
+
+    private void visitRowType(RowType type, String name, List<String> parent)
+    {
+        parent = ImmutableList.<String>builder().addAll(parent).add(name).build();
+        for (RowType.Field field : type.getFields()) {
+            com.google.common.base.Preconditions.checkArgument(field.getName().isPresent(), "field in struct type doesn't have name");
```

Comment:
Static import `checkArgument`

Reply:
Changed.  I have to remember to check code that I cut and paste, because the original might not be conformant.